src/QsCompiler/Transformations/ContentLifting.cs

```diff
@@ -0,0 +1,545 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+using Microsoft.Quantum.QsCompiler.Transformations.SearchAndReplace;
+
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.ContentLifting
+{
+    using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
+    using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+    using TypeArgsResolution = ImmutableArray<Tuple<QsQualifiedName, NonNullable<string>, ResolvedType>>;
+
+    /// <summary>
+    /// Transformation handling the task of lifting the contents of code blocks into generated operations.
+    /// The transformation provides validation to see if any given block can safely be lifted into its own operation.
+    /// Validation requirements are that there are no return statements and that there are no set statements
+    /// on mutables declared outside the block. Setting mutables declared inside the block is valid.
+    /// A block can be checked by setting the SharedState.IsValidScope to true before traversing the scope,
```

Comment:
We don't necessarily want it to be use at the beginning of OnScope. Where this gets used is dependent on the derived behavior of the more specific lift transformation. For example, in classical control's lift transformation, this gets used in very context-specific logic in the OnConditionalStatement function.

Reply:
It would be nice to have the validation logic all encapsulated in the base LiftContent transformation's LiftBody function. But for now this validation logic is too entangled with the context-specific logic in the LiftConditionBlocks transformation.